kernel/sched.c

```diff
@@ -1338,45 +1338,64 @@ struct k_thread *z_priq_rb_best(struct _priq_rb *pq)
 }
 
 #ifdef CONFIG_SCHED_MULTIQ
-# if (K_LOWEST_THREAD_PRIO - K_HIGHEST_THREAD_PRIO) > 31
-# error Too many priorities for multiqueue scheduler (max 32)
-# endif
+
+struct prio_info {
+	uint8_t offset_prio;
+	uint8_t idx;
+	uint8_t bit;
+};
+
+static inline struct prio_info get_prio_info(int8_t old_prio)
+{
+	struct prio_info ret;
+
+	ret.offset_prio = old_prio - K_HIGHEST_THREAD_PRIO;
+	ret.idx = ret.offset_prio / NBITS;
+	ret.bit = ret.offset_prio & (NBITS - 1);
+
+	return ret;
+}
 
 static ALWAYS_INLINE void z_priq_mq_add(struct _priq_mq *pq,
 					struct k_thread *thread)
 {
-	int priority_bit = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	struct prio_info pos = get_prio_info(thread->base.prio);
 
-	sys_dlist_append(&pq->queues[priority_bit], &thread->base.qnode_dlist);
-	pq->bitmask |= BIT(priority_bit);
+	sys_dlist_append(&pq->queues[pos.offset_prio], &thread->base.qnode_dlist);
+	pq->bitmask[pos.idx] |= BIT(pos.bit);
 }
 
 static ALWAYS_INLINE void z_priq_mq_remove(struct _priq_mq *pq,
 					   struct k_thread *thread)
 {
-	int priority_bit = thread->base.prio - K_HIGHEST_THREAD_PRIO;
+	struct prio_info pos = get_prio_info(thread->base.prio);
 
 	sys_dlist_remove(&thread->base.qnode_dlist);
-	if (sys_dlist_is_empty(&pq->queues[priority_bit])) {
-		pq->bitmask &= ~BIT(priority_bit);
+	if (sys_dlist_is_empty(&pq->queues[pos.offset_prio])) {
+		pq->bitmask[pos.idx] &= ~BIT(pos.bit);
 	}
 }
 #endif
 
 struct k_thread *z_priq_mq_best(struct _priq_mq *pq)
 {
-	if (!pq->bitmask) {
-		return NULL;
-	}
-
-	struct k_thread *thread = NULL;
-	sys_dlist_t *l = &pq->queues[__builtin_ctz(pq->bitmask)];
-	sys_dnode_t *n = sys_dlist_peek_head(l);
+	for (int i = 0; i < NUM_PRIO; ++i) {
+		if (!pq->bitmask[i]) {
+			continue;
+		}
```

Comment:
`i < NUM_PRIO` is not the right termination condition (as it would in theory allow for accessing beyond the bounds of pq->bitmask[]--in practice though, I doubt that would happen). Better, I think, to use `i < PRIQ_BITMAP_SIZE`.

Reply:
Fixed.